blob/blob.go

```diff
@@ -16,42 +19,70 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+//nolint:unused
 var errEmptyShares = errors.New("empty shares")
 
-// The Proof is a set of nmt proofs that can be verified only through
-// the included method (due to limitation of the nmt https://github.com/celestiaorg/nmt/issues/218).
-// Proof proves the WHOLE namespaced data to the row roots.
-// TODO (@vgonkivs): rework `Proof` in order to prove a particular blob.
-// https://github.com/celestiaorg/celestia-node/issues/2303
-type Proof []*nmt.Proof
-
-func (p Proof) Len() int { return len(p) }
+// Proof constructs the proof of a blob to the data root.
+type Proof struct {
+	// ShareToRowRootProof the proofs of the shares to the row roots they belong to.
+	// If the blob spans across multiple rows, then this will contain multiple proofs.
+	ShareToRowRootProof []*tmproto.NMTProof
+	// RowToDataRootProof the proofs of the row roots containing the blob shares
+	// to the data root.
+	RowToDataRootProof coretypes.RowProof
```

Comment:
Q: why do we call DataRoot if inside header it is called `DataHash`?

Reply:
That's an existential question. From the dawn of existence, the name `DataRoot` has echoed through the ages, a timeless whisper woven into the fabric of eternity.

That's a theory vs implementation issue... In the fraud proofs paper, it's referred to as `DataRoot`. In tendermint, it's called `DataHash`, and AFAIK we never really discussed choosing one over the other.